app/vminsert/netstorage/insert_ctx.go

```diff
@@ -134,26 +154,30 @@ func (ctx *InsertCtx) WriteDataPointExt(at *auth.Token, storageNodeIdx int, metr
 // FlushBufs flushes ctx bufs to remote storage nodes.
 func (ctx *InsertCtx) FlushBufs() error {
 	var firstErr error
-	for i := range ctx.bufRowss {
-		br := &ctx.bufRowss[i]
-		if len(br.buf) == 0 {
-			continue
-		}
-		if err := br.pushTo(storageNodes[i]); err != nil && firstErr == nil {
-			firstErr = err
+	for rgID, rg := range replicationGroups {
+		bufRowss := ctx.groupBufRowss[rgID]
+		for i := range bufRowss {
+			br := &bufRowss[i]
+			if len(br.buf) == 0 {
+				continue
+			}
+			if err := br.pushTo(rg.storageNodes[i]); err != nil && firstErr == nil {
+				firstErr = err
+			}
 		}
 	}
 	return firstErr
 }
 
-// GetStorageNodeIdx returns storage node index for the given at and labels.
-//
-// The returned index must be passed to WriteDataPoint.
-func (ctx *InsertCtx) GetStorageNodeIdx(at *auth.Token, labels []prompb.Label) int {
-	if len(storageNodes) == 1 {
-		// Fast path - only a single storage node.
-		return 0
-	}
+// StorageNodeGroupID keeps the integer id of a node and it's group
+type StorageNodeGroupID struct {
```

Comment:
This type isn't needed when using a simple loop over `replicationGroups` during data write as described above